### ISSUE TITLE
inconsistent policy map entries for CIDR policy

### ISSUE THREAD

**Issue by sypakine** [cilium/cilium#23750]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

tl;dr: host-firewall policies produce inconsistent policy maps for CIDR entries.

Context

Using host firewall (ccnp w/ node selector) to restrict ingress to control-plane nodes. 

```
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "control-plane-nodes-allow-node-ingress"
spec:
  nodeSelector:
    matchExpressions:
    - key: node-role.kubernetes.io/control-plane
      operator: Exists
  ingress:
  - fromEntities:
    - remote-node
  - ...
```

During node upgrade, prior to the node joining the cluster, the IP address of the node being created no longer maps to `remote-node` identity and ingress is denied.

To circumvent this, added the `/32` CIDRs of the nodes to the policy:

```
...
spec:
  nodeSelector:
    matchExpressions:
    - key: node-role.kubernetes.io/control-plane
      operator: Exists
  ingress:
  - fromEntities:
    - remote-node
  - fromCIDR:
    - 192.168.8.2/32
    - 192.168.8.3/32
    - 192.168.8.4/32
  - ...
```

---

Problem

When a CIDR IP maps to a reserved identity, the resulting identity lookup can result in inconsistent policy map.

1. CIDR entries are allocated to determine whether to create a new identity. To do so, Labels are created that always has (at minimum) 2 parts (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/ipcache/cidr.go#L37):
  - `cidr:${IP}/${CIDR}` (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/labels/cidr/cidr.go#L87)
  - `reserved:world` (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/labels/cidr/cidr.go#L104)

2. The Labels are then merged with the Labels retrieved by looking up the IP in the ipcache metadata (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/ipcache/cidr.go#L60). This is where the problem starts.

3. The labels are allocated an identity, which looks up whether the identity maps to a reserved identity (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/identity/cache/allocator.go#L352).
  - in the above context - if this policy is created when the nodes are present - this would add the labels:
    - `reserved:remote-node`
    - `reserved:kube-apiserver`
    - (and if the CIDR maps to the same host `reserved:host`) 

5. This loops over the labels, and has logic to handle reserved labels (ref: https://github.com/cilium/cilium/blob/v1.12/pkg/identity/identity.go#L224). As Labels is a map, the order for which they are processed is random.

For the context above, one of our Labels would at minimum look like:

- `cidr:192.168.8.2/32`
- `reserved: world`
- `reserved:remote-node`
- `reserved:kube-apiserver`

The logic does not have a case for `reserved: world`, so if `reserved: world` is processed prior to `reserved:kube-apiserver` (or `reserved:host`, if present), the resulting identity differs. In the case `reserved:kube-apiserver` is processed first, the reserved identity value (`7`) is reused. If `reserved: world` is processed a new identity value is generated.

---

For our use-case, declaring a CIDR should always insert an entry in the policy map with only the `cidr` and `world` labels. This policy won't be matched against as long as the Node entity maps to `remote-node` or `kube-apiserver`, and will only come into play when the node has been removed from the cluster and a new ipcache entry (with the new identity for the IP address) is generated .

A workaround is to use a CIDR range (/31+), but even this is not fool-proof as the network IP can still map to a reserved identity:

```
Allow    Ingress     cidr:192.168.8.4/31           ANY          NONE         0         0
                     reserved:remote-node
                     reserved:world
```

--- 

I have tested with a custom v1.12 image version which skips the label merge for CIDRs, and this consistently produces the correct behavior. However, I am not certain of the downstream effects of implementing this change

### Cilium Version

`v1.12`

### Kernel Version

`Linux kind-worker2 5.19.11-1rodete1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 5.19.11-1rodete1 (2022-10-31) x86_64 x86_64 x86_64 GNU/Linux`

### Kubernetes Version

`$k version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.24.2
Kustomize Version: v4.5.4
Server Version: v1.21.1

### Sysdump

Can provide a sys dump if required.

### Anything else?

Reproduction:

Setup:

```
cat > kind-config.yaml <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: control-plane
- role: control-plane
- role: worker
- role: worker
- role: worker
networking:
  disableDefaultCNI: true
EOF

CILIUM_CLI_VERSION=$(curl -s https://raw.githubusercontent.com/cilium/cilium-cli/master/stable.txt)
CLI_ARCH=amd64
if [ "$(uname -m)" = "aarch64" ]; then CLI_ARCH=arm64; fi
curl -L --fail --remote-name-all https://github.com/cilium/cilium-cli/releases/download/${CILIUM_CLI_VERSION}/cilium-linux-${CLI_ARCH}.tar.gz{,.sha256sum}
sha256sum --check cilium-linux-${CLI_ARCH}.tar.gz.sha256sum
sudo tar xzvfC cilium-linux-${CLI_ARCH}.tar.gz /usr/local/bin
rm cilium-linux-${CLI_ARCH}.tar.gz{,.sha256sum}
 
alias k=kubectl && \
kind create cluster --config=kind-config.yaml && \
k config set-context --current --namespace=kube-system && \
cilium install && \
cilium status --wait && \
k patch cm cilium-config --patch '
data:
  enable-host-firewall: "true"
  debug: "true"
  debug-verbose: "true"
' && \
k rollout restart ds/cilium && k rollout status ds/cilium && \
k apply -f - <<EOF
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "allow-intra-control-plane-node"
spec:
  nodeSelector:
    matchExpressions:
    - key: node-role.kubernetes.io/control-plane
      operator: Exists
  ingress:
  - fromEntities:
    - remote-node
  # allow external access
  - fromCIDR:
    - 10.0.0.0/8
  - fromCIDR:
    - 192.168.8.2/32
    - 192.168.8.3/32
    - 192.168.8.4/32
    - 192.168.8.5/32
    - 192.168.8.6/32
    - 192.168.8.7/32
    - 192.168.8.8/32
EOF
```

Observe:

```
k get pod -l k8s-app=cilium -ojsonpath='{range .items[*]}{@.metadata.name}{"\n"}{end}' | xargs -I {} -i sh -c "echo {}; kubectl exec -i {} -c cilium-agent -n kube-system -- cilium bpf policy get --all | grep -E 'cidr:192.168.8.[2-8]' -A 3" | grep -v Egress
 
k get pod -l k8s-app=cilium -ojsonpath='{range .items[*]}{@.metadata.name}{"\n"}{end}' | xargs -I {} -i sh -c "echo {}; kubectl exec -i {} -c cilium-agent -n kube-system -- cilium bpf ipcache list | grep -E '^192.168.8.[2-8]'"
 
k get pod -l k8s-app=cilium -ojsonpath='{range .items[*]}{@.metadata.name}{"\n"}{end}' | xargs -I {} -i sh -c "echo {}; kubectl exec -i {} -c cilium-agent -n kube-system -- cilium identity list | grep -E '192.168.8.[2-8]' -A3"
```

(in rare cases, may be necessary to delete/recreate the policy)

Can see that the agents do not have a consistent view of the policy

e.g. 

`cilium-wzpw5` in my reproduction is missing `192.168.8.6/32` from the policy:
```
# cilium bpf policy get --all | grep -E 'cidr:192.168.8.[2-8]' -A 3" 
Allow    Ingress     cidr:192.168.8.4/32           ANY          NONE         0        0
                     reserved:host
                     reserved:world
Allow    Ingress     cidr:192.168.8.5/32           ANY          NONE         0        0
                     reserved:kube-apiserver
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.3/32           ANY          NONE         0        0
                     reserved:kube-apiserver
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.7/32           ANY          NONE         0        0
                     reserved:kube-apiserver
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.2/32           ANY          NONE         169967   335
                     reserved:world
Allow    Ingress     cidr:192.168.8.8/32           ANY          NONE         0        0
                     reserved:remote-node
                     reserved:world
```

and results in denying ingress from the IP:

```
Policy verdict log: flow 0x0 local EP ID 2615, remote ID world, proto 6, ingress, action deny, match none, 192.168.8.6:2883 -> 192.168.8.4:0 tcp SYN
```

vs `cilium-npkcf` which has a reference for all:
```

# cilium bpf policy get --all | grep -E 'cidr:192.168.8.[2-8]' -A 3" 
Allow    Ingress     cidr:192.168.8.3/32           ANY          NONE         0          0
                     reserved:host
                     reserved:kube-apiserver
                     reserved:world
Allow    Ingress     cidr:192.168.8.4/32           ANY          NONE         0          0
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.6/32           ANY          NONE         2657       28
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.8/32           ANY          NONE         0          0
                     reserved:remote-node
                     reserved:world
Allow    Ingress     cidr:192.168.8.2/32           ANY          NONE         630469     3100
                     reserved:world
Allow    Ingress     cidr:192.168.8.7/32           ANY          NONE         0          0
                     reserved:kube-apiserver
                     reserved:remote-node
                     reserved:world
```

When I delete the node associated with internal address `192.168.8.6`, `cilium-wzpw5` begins denying all ingress traffic from `192.168.8.6`

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christarazi** [comment]
Thanks for filing this issue. 

This is a bug and unfortunately, something that was known to us for a while, see #17962.

>  Handle ToCIDR policy followed by kube-apiserver policy case in terms of identity allocation, ipcache entry fixup, etc.

**Comment by christarazi** [comment]
@sypakine Could you try https://github.com/cilium/cilium/pull/23764 and see if it fixes your issue?
